src/e3/json.py

```diff
@@ -6,16 +6,71 @@
 
 import e3.error
 
+from abc import ABC, abstractmethod
 from typing import TYPE_CHECKING
 
 if TYPE_CHECKING:
-    from typing import Any
+    from typing import Any, TypeVar
+
+    JsonDataSelf = TypeVar("JsonDataSelf", bound="JsonData")
 
 
 class JsonError(e3.error.E3Error):
     pass
 
 
+class JsonData(ABC):
+    """An object to represent JSON data content."""
+
+    @abstractmethod
+    def as_dict(self) -> dict[str, object]:
+        """Return the dict representation of this JSON data object."""
+        ...
+
+    def __eq__(self, other: object) -> bool:
+        """Check if this JSON data is identical to *other*.
+
+        :param other: The object to compare this JSON data with.
+
+        :return: A :class:`python:bool` set to **True** if both JSON data are identical,
+            **False** if they are not, or if *other* is not a CPE.
+        """  # noqa RST304
+        if isinstance(other, self.__class__):
+            return self.as_json() == other.as_json()
+        return False
+
+    def as_json(self) -> str:
+        """Return a JSON string representing this JSON data.
+
+        .. seealso:: :func:`python:json.dump`
+        """  # noqa RST304
+        return json.dumps(self.as_dict(), sort_keys=True)
+
+    @classmethod
+    def from_dict(cls: type[JsonDataSelf], obj: dict) -> JsonDataSelf:
+        """Load a dictionary as a JSON data object.
+
+        :param obj: The dictionary to initialize the JSON data object with.
+
+        :return: A new :class:`JsonData` object initialized with
+            values from the input dictionary.
+        """  # noqa RST304
+        return cls(**obj)
+
+    @classmethod
+    def from_json(cls: type[JsonDataSelf], content: str) -> JsonDataSelf:
+        """Load a JSON string as a JSON data object.
+
+        :param content: The JSON string to initialize the JSON data object with.
+
+        :return: A new :class:`JsonData` object initialized with
+            values from the input dictionary.
+
+        .. seealso:: :meth:`from_dict`
+        """  # noqa RST304
+        return cls.from_dict(json.loads(content))
```

Comment:
A JSON string does not always represent a dict, it can be a list or even a simple string.
If this class is meant to work only with dicts maybe  the "dict" constraint should appear
in the class name and/or method names (`from_json` -> `from_json_dict` for example).

Reply:
I'd rather update the doc string to state that _content must be loadable through a call to `json.loads()`_

Reply:
``` 
>>> json.loads('["a_json"]')
['a_json']
```

Reply:
In any case, this is more a doc issue. Naming it `from_json_dict()` will make (at least me) think it loads a `dict`, not a string.

Reply:
In addition to doc maybe also raise a specific exception when `json.loads` does not return a dict with a clear message indicating that only dictionaries are supported.